Job creation fails during fabric8:apply with error: forbidden

<!-- Please fill out the following information to help us in analyzing the issue, but feel free to skip it if you don't have the information at hand or if it does not apply. Please remove every section which does not apply to the issue. -->
### Description
When using `mvn fabric8:resource fabric8:apply` in the given sample project i get the following error:
```
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building abcd 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- fabric8-maven-plugin:4.1.0:resource (default-cli) @ abcd ---
[INFO] F8: Using docker image name of namespace: myproject
[INFO] F8: using resource templates from C:\Users\vlutz\Projects\openshift-job\fabric8
[INFO] F8: fmp-revision-history: Adding revision history limit to 2
[INFO] F8: validating C:\Users\vlutz\Projects\openshift-job\target\classes\META-INF\fabric8\kubernetes\pi-job.yml resource
[INFO] F8: using resource templates from C:\Users\vlutz\Projects\openshift-job\fabric8
[INFO] F8: fmp-revision-history: Adding revision history limit to 2
[INFO] F8: validating C:\Users\vlutz\Projects\openshift-job\target\classes\META-INF\fabric8\openshift\pi-job.yml resource
[INFO] 
[INFO] --- fabric8-maven-plugin:4.1.0:apply (default-cli) @ abcd ---
[INFO] F8: Using OpenShift at https://10.0.75.2:8443/ in namespace myproject with manifest C:\Users\vlutz\Projects\openshift-job\target\classes\META-INF\fabric8\openshift.yml 
[INFO] OpenShift platform detected
[INFO] F8: Using project: myproject
[INFO] F8: Applying Job pi from openshift.yml
[ERROR] F8: Failed to create Job from openshift.yml. io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.0.75.2:8443/apis/batch/v1/namespaces/myproject/jobs/pi. Message: Forbidden! User developer/10-0-75-2:8443 doesn't have permission. jobs.extensions "pi" is forbidden: User "developer" cannot get jobs.extensions in the namespace "myproject": User "developer" cannot get jobs.extensions in project "myproject".: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.0.75.2:8443/apis/batch/v1/namespaces/myproject/jobs/pi. Message: Forbidden! User developer/10-0-75-2:8443 doesn't have permission. jobs.extensions "pi" is forbidden: User "developer" cannot get jobs.extensions in the namespace "myproject": User "developer" cannot get jobs.extensions in project "myproject".
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 7.894 s
[INFO] Finished at: 2019-06-05T10:49:15+02:00
[INFO] Final Memory: 22M/77M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.fabric8:fabric8-maven-plugin:4.1.0:apply (default-cli) on project abcd: Failed to create Job from openshift.yml. io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.0.75.2:8443/apis/batch/v1/namespaces/myproject/jobs/pi. Message: Forbidden! User developer/10-0-75-2:8443 doesn't have permission. jobs.extensions "pi" is forbidden: User "developer" cannot get jobs.extensions in the namespace "myproject": User "developer" cannot get jobs.extensions in project "myproject". -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```
But with curl i have no permission problem
```
curl -k -H "Authorization: Bearer qKqaYkIFumcwX7CtZ711CcUt_ILTFpijEqQrWfLtCQE" -H 'Accept: application/json' https ://10.0.75.2:8443/apis/batch/v1/namespaces/myproject/jobs
{
  "kind": "JobList",
  "apiVersion": "batch/v1",
  "metadata": {
    "selfLink": "/apis/batch/v1/namespaces/myproject/jobs",
    "resourceVersion": "7131"
  },
  "items": []
```
```
 curl -k -H "Authorization: Bearer qKqaYkIFumcwX7CtZ711CcUt_ILTFpijEqQrWfLtCQE" -H 'Accept: application/json' https ://10.0.75.2:8443/apis/batch/v1/namespaces/myproject/jobs/pi
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "jobs.batch \"pi\" not found",
  "reason": "NotFound",
  "details": {
    "name": "pi",
    "group": "batch",
    "kind": "jobs"
  },
  "code": 404
```
I alos can create the job with the yaml file over the openshift-cli
```
oc create -f job.yaml
job "pi" created
```
### Info

* f-m-p version : 4.1.0
* Maven version (`mvn -v`) : 3.3.9
* Kubernetes / OpenShift setup and version : v3.9.0+191fece
* Sample project : https://github.com/ValentinLutz/openshift-job.git

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@ After this we will switch probably to real [Semantic Versioning 2.0.0](http://se
 * Fix #1636: Added support to choose configMap name
 * Fix #1630: Fix CronJob Version
 * Fix #1591: Add support for custom resources creation via resource fragments
+* Fix #1648: Job creation fails during fabric8:apply with error: forbidden
 
 ### 4.1.0 (16-02-2019)
 * Fix 1578: Fmp is generating ImageChange triggers in the DeploymentConfig

--- a/core/src/main/java/io/fabric8/maven/core/service/ApplyService.java
+++ b/core/src/main/java/io/fabric8/maven/core/service/ApplyService.java
@@ -33,10 +33,12 @@ import io.fabric8.kubernetes.api.model.apps.DaemonSet;
 import io.fabric8.kubernetes.api.model.apps.Deployment;
 import io.fabric8.kubernetes.api.model.apps.ReplicaSet;
 import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.kubernetes.api.model.batch.Job;
 import io.fabric8.kubernetes.api.model.extensions.Ingress;
 import io.fabric8.kubernetes.api.model.rbac.Role;
 import io.fabric8.kubernetes.api.model.rbac.RoleBinding;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.dsl.MixedOperation;
 import io.fabric8.kubernetes.client.dsl.Resource;
 import io.fabric8.kubernetes.client.dsl.base.CustomResourceDefinitionContext;
@@ -65,6 +67,7 @@ import org.apache.commons.lang3.StringUtils;
 
 import java.io.File;
 import java.io.IOException;
+import java.net.HttpURLConnection;
 import java.nio.charset.Charset;
 import java.util.ArrayList;
 import java.util.List;
@@ -184,6 +187,8 @@ public class ApplyService {
             applyPersistentVolumeClaim((PersistentVolumeClaim) dto, sourceName);
         }else if (dto instanceof CustomResourceDefinition) {
             applyCustomResourceDefinition((CustomResourceDefinition) dto, sourceName);
+        } else if (dto instanceof Job) {
+            applyJob((Job) dto, sourceName);
         } else if (dto instanceof HasMetadata) {
             HasMetadata entity = (HasMetadata) dto;
             try {
@@ -1209,6 +1214,37 @@ public class ApplyService {
         } catch (Exception e) {
             onApplyError("Failed to create Pod from " + sourceName + ". " + e + ". " + pod, e);
         }
+    }
+
+    protected void applyJob(Job job, String sourceName) {
+        String namespace = getNamespace();
+        String id = getName(job);
+        Objects.requireNonNull(id, "No name for " + job + " " + sourceName);
+        if (isServicesOnlyMode()) {
+            log.debug("Only processing Services right now so ignoring Job: " + namespace + ":" + id);
+            return;
+        }
+        // Not using createOrReplace() here (https://github.com/fabric8io/kubernetes-client/issues/1586)
+        try {
+            doCreateJob(job, namespace, sourceName);
+        } catch (KubernetesClientException exception) {
+            if(exception.getStatus().getCode().equals(HttpURLConnection.HTTP_CONFLICT)) {
+                Job old = kubernetesClient.batch().jobs().inNamespace(namespace).withName(id).get();
+                Job updatedJob = patchService.compareAndPatchEntity(namespace, job, old);
+                log.info("Updated Job: " + old.getMetadata().getName());
+                return;
+            }
+            onApplyError("Failed to apply Job from " + job.getMetadata().getName(), exception);
+        }
+    }
+
+    public void doCreateJob(Job job, String namespace, String sourceName) throws KubernetesClientException {
+        if (StringUtils.isNotBlank(namespace)) {
+            kubernetesClient.batch().jobs().inNamespace(namespace).create(job);
+        } else {
+            kubernetesClient.batch().jobs().inNamespace(getNamespace()).create(job);
+        }
+        log.info("Creating a Job from " + sourceName + " namespace " + namespace + " name " + getName(job));
     }
 
     public String getNamespace() {

--- a/core/src/main/java/io/fabric8/maven/core/service/PatchService.java
+++ b/core/src/main/java/io/fabric8/maven/core/service/PatchService.java
@@ -28,6 +28,8 @@ import io.fabric8.kubernetes.api.model.Secret;
 import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.api.model.apiextensions.CustomResourceDefinition;
 import io.fabric8.kubernetes.api.model.apiextensions.DoneableCustomResourceDefinition;
+import io.fabric8.kubernetes.api.model.batch.DoneableJob;
+import io.fabric8.kubernetes.api.model.batch.Job;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.maven.core.util.kubernetes.OpenshiftHelper;
 import io.fabric8.maven.core.util.kubernetes.UserConfigurationCompare;
@@ -74,6 +76,7 @@ public class PatchService {
         patchers.put("Secret", secretPatcher());
         patchers.put("PersistentVolumeClaim", pvcPatcher());
         patchers.put("CustomResourceDefinition", crdPatcher());
+        patchers.put("Job", jobPatcher());
     }
 
 
@@ -225,6 +228,30 @@ public class PatchService {
             if (!UserConfigurationCompare.configEqual(newObj.getSpec(), oldObj.getSpec())) {
                 entity.withSpec(newObj.getSpec());
             }
+            return entity.done();
+        };
+    }
+
+    private static EntityPatcher<Job> jobPatcher() {
+        return (KubernetesClient client, String namespace, Job newObj, Job oldObj) -> {
+            if (UserConfigurationCompare.configEqual(newObj, oldObj)) {
+                return oldObj;
+            }
+
+            DoneableJob entity = client.batch().jobs().withName(oldObj.getMetadata().getName()).edit();
+
+            if (!UserConfigurationCompare.configEqual(newObj.getMetadata(), oldObj.getMetadata())) {
+                entity.withMetadata(newObj.getMetadata());
+            }
+
+            if (!UserConfigurationCompare.configEqual(newObj.getSpec().getSelector(), oldObj.getSpec().getSelector())) {
+                entity.editSpec().withSelector(newObj.getSpec().getSelector());
+            }
+
+            if (!UserConfigurationCompare.configEqual(newObj.getSpec().getTemplate(), oldObj.getSpec().getSelector())) {
+                entity.editSpec().withTemplate(newObj.getSpec().getTemplate());
+            }
+
             return entity.done();
         };
     }

--- a/enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ProjectLabelEnricher.java
+++ b/enricher/standard/src/main/java/io/fabric8/maven/enricher/standard/ProjectLabelEnricher.java
@@ -22,7 +22,6 @@ import io.fabric8.kubernetes.api.model.ServiceBuilder;
 import io.fabric8.kubernetes.api.model.apps.DaemonSetBuilder;
 import io.fabric8.kubernetes.api.model.apps.DeploymentBuilder;
 import io.fabric8.kubernetes.api.model.apps.StatefulSetBuilder;
-import io.fabric8.kubernetes.api.model.batch.JobBuilder;
 import io.fabric8.maven.core.config.PlatformMode;
 import io.fabric8.maven.core.util.Configs;
 import io.fabric8.maven.core.util.MapUtil;
@@ -130,19 +129,6 @@ public class ProjectLabelEnricher extends BaseEnricher {
                 builder.editOrNewSpec().editOrNewSelector().withMatchLabels(selectors).endSelector().endSpec();
             }
         });
-
-        builder.accept(new TypedVisitor<JobBuilder>() {
-            @Override
-            public void visit(JobBuilder builder) {
-                Map<String, String> selectors = new HashMap<>();
-                if(builder.buildSpec() != null && builder.buildSpec().getSelector() != null && builder.buildSpec().getSelector().getMatchLabels() != null) {
-                    selectors.putAll(builder.buildSpec().getSelector().getMatchLabels());
-                }
-                MapUtil.mergeIfAbsent(selectors, createLabels());
-                builder.editOrNewSpec().editOrNewSelector().withMatchLabels(selectors).endSelector().endSpec();
-            }
-        });
-
     }
 
     @Override

